### PR TITLE
[ᚬrc/v0.14.0] [BREAKCHANGE] feat: display miner worker status

### PR DESCRIPTION
**Pull request by quake**
BREAKCHANGE: config file `ckb-miner.toml` changed

- feat: display miner worker status
- refactor: add miner worker trait, prepare for `Cuckoo` cpu and gpu miner plugin

tty

### CODE REVIEW

File: miner/src/worker/mod.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    }
+}
+
+pub fn start_worker(
+    pow: Arc<dyn PowEngine>,
+    config: &WorkerConfig,
+    seal_tx: Sender<(H256, Seal)>,
+) -> WorkerController {
+    let mp = MultiProgress::new();
+    let spinner_style = ProgressStyle::default_spinner()
```
Comment (doitian) [inline miner/src/worker/mod.rs]: How it resolves the conflicts with logger?

File: docs/ckb-core-dev.md [hunk 2]
```diff
@@ -60,9 +60,13 @@ cycle_length = 12
...
 func = "Dummy"
+```
+
+and don't forget to modify `ckb-miner.toml` miner workers section:
 
-# Delay offset (in milliseconds)
-[pow.params.delay]
-type = "constant"
-value = 5000
+```
```
[the remark this answers is not in the document]
Comment (quake) [reply docs/ckb-core-dev.md]: it's related to dummy miner only, no need to set these value in chain config.

File: miner/src/worker/mod.rs (hunk 1 above)
Comment (quake) [reply miner/src/worker/mod.rs]: progress bar log to stderr, no conflict with the logger.

File: docs/ckb-core-dev.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (doitian) [reply docs/ckb-core-dev.md]: Got it. My point is that it is easier to switch PoW if we can only edit it in one file. Maybe we can provide params for all PoW engines in `ckb-miner.yml`, and only the params for current PoW will be used.

File: miner/src/worker/mod.rs (hunk 1 above)
Comment (doitian) [reply miner/src/worker/mod.rs]: Some may want to run miner as a daemon. The progress should be disabled if so. Let's add `StandardError=null` under `[Service]` in devtools/init/linux-systemd/ckb-miner.service

**Comment by quake** [comment]
> Disable stderr in devtools/init/linux-systemd/ckb-miner.service

Done, please review again, thanks.

File: resource/ckb-miner.toml [hunk 3]
```diff
@@ -32,11 +30,22 @@ dsn = "" # {{
...
 # _ => rpc_url = "http://127.0.0.1:{rpc_port}/"
 # }}
 block_on_submit = true
 
 # block template polling interval in milliseconds
 poll_interval = 1000
+
+[[miner.workers]]
+worker_type = "CuckooSimple"
+threads     = "1"
```
Comment (TheWaWaR) [inline resource/ckb-miner.toml]: Why not integer?

**Comment by TheWaWaR** [comment]
Maybe a config to turn off status output?

File: resource/ckb-miner.toml [hunk 4]
```diff
@@ -32,11 +30,22 @@ dsn = "" # {{
...
 
-[miner]
+[miner.client]
 rpc_url = "http://127.0.0.1:8114/" # {{
 # _ => rpc_url = "http://127.0.0.1:{rpc_port}/"
 # }}
 block_on_submit = true
 
 # block template polling interval in milliseconds
 poll_interval = 1000
```
Comment (TheWaWaR) [inline resource/ckb-miner.toml]: Maybe a config to turn off status output?

File: resource/ckb-miner.toml (hunk 3 above)
Comment (quake) [reply resource/ckb-miner.toml]: We need to support `bool` or `float` here, for example: 
```
[[miner.workers]]
worker_type = "CuckooCpuMean"
trim        = "true"

[[miner.workers]]
worker_type = "Dummy"
delay_type  = "Normal"
mean        =  "5000"
std_dev     =  "32.5"
```

File: resource/ckb-miner.toml (hunk 4 above)
Comment (quake) [reply resource/ckb-miner.toml]: > Maybe a config to turn off status output?

I'm considering adding more status to the console output, let me add this configuration parameter with more status in another PR.

File: docs/ckb-core-dev.md [hunk 5]
```diff
@@ -60,9 +60,13 @@ cycle_length = 12
...
+[[miner.workers]]
+worker_type = "Dummy"
```
Comment (zhangsoledad) [inline docs/ckb-core-dev.md]: I think different types do not allow to exist at the same time, pow type should be global

**Comment by yangby-cryptape** [comment]
Last week, I asked a question in chat: could the numbers in the result be equal the `max_edge`. When I was reviewing this PR, I felt confused again.

- In our implementation:

  We didn't allow that when generating the edges: https://github.com/nervosnetwork/ckb/blob/a1db895e57c6a3ced78b81d26af0c5815c26eab1/pow/src/cuckoo.rs#L273-L281

  But, we allowed that when doing verification: https://github.com/nervosnetwork/ckb/blob/a1db895e57c6a3ced78b81d26af0c5815c26eab1/pow/src/cuckoo.rs#L185-L187

- And, I thought the official implementation do the same logic, too.

File: docs/ckb-core-dev.md (hunk 5 above)
Comment (quake) [reply docs/ckb-core-dev.md]: pow type is global, but the user may need to configure different worker type, for example, running 1 miner on CPU and two miners on different GPUs:

```
[[miner.workers]]
worker_type = "CuckooCpu"
threads     = "2"

[[miner.workers]]
worker_type = "CuckooGpuLean"
device_id   = "..."

[[miner.workers]]
worker_type = "CuckooGpuMean"
device_id   = "..."
```

File: resource/ckb-miner.toml (hunk 3 above)
Comment (doitian) [reply resource/ckb-miner.toml]: You can use `toml::Value` as the hashmap value.

Comment (doitian) [reply resource/ckb-miner.toml]: Another choise: use internal tagged enum

```rust
#[derive(Serialize, Deserialize)]
#[serde(tag = "worker_type")]
enum WorkerConfig {
    CuckooCpuMean { trim: bool },
    Dummy { delay_type: String, mean: u32, std_dev: f64 },
}
```

File: docs/ckb-core-dev.md (hunk 5 above)
Comment (zhangsoledad) [reply docs/ckb-core-dev.md]: I don't think such `Dummy` here is served for your purpose. It's different level

File: resource/ckb-miner.toml (hunk 3 above)
Comment (quake) [reply resource/ckb-miner.toml]: @doitian @TheWaWaR I pushed a new commit https://github.com/nervosnetwork/ckb/pull/961/commits/4059a26365b3cda571ebc7d5cfdae9ca62662d0c to resolve this issue, please review again, thanks

File: docs/ckb-core-dev.md (hunk 5 above)
Comment (quake) [reply docs/ckb-core-dev.md]: @zhangsoledad I'm going to stick on this commit: https://github.com/nervosnetwork/ckb/commit/4059a26365b3cda571ebc7d5cfdae9ca62662d0c after fighting with `toml::Value`, please review again, thanks.

File: util/app-config/src/app_config.rs [hunk 6]
```diff
@@ -401,45 +401,7 @@ mod tests {
                 miner_config.chain.spec,
                 PathBuf::from("specs/integration.toml")
             );
-            assert_eq!(miner_config.miner.rpc_url, "http://127.0.0.1:7000/");
-        }
-    }
-
-    #[cfg(all(unix, target_pointer_width = "64"))]
```
Comment (doitian) [inline util/app-config/src/app_config.rs]: Wrong merge?

Comment (doitian) [reply util/app-config/src/app_config.rs]: The test is added by @xxuejie when porting to Windows.

Comment (quake) [reply util/app-config/src/app_config.rs]: sorry, resolved auto rebase error, please review again.

**Pull request merged**